cmd/gaia/cli_test/cli_test.go

```diff
@@ -836,6 +836,45 @@ func TestGaiaCLIEncode(t *testing.T) {
 	require.Equal(t, "deadbeef", decodedTx.Memo)
 }
 
+func TestGaiaCLIDecode(t *testing.T) {
+	t.Parallel()
+	f := InitFixtures(t)
+
+	// start gaiad server
+	proc := f.GDStart()
+	defer proc.Stop(false)
+
+	// Build a testing transaction and write it to disk
+	barAddr := f.KeyAddress(keyBar)
+	keyAddr := f.KeyAddress(keyFoo)
+
+	sendTokens := sdk.TokensFromTendermintPower(10)
+	success, stdout, stderr := f.TxSend(keyAddr.String(), barAddr, sdk.NewCoin(denom, sendTokens), "--generate-only", "--memo", "deadbeef")
+	require.True(t, success)
+	require.Empty(t, stderr)
+
+	// Write it to disk
+	jsonTxFile := WriteToNewTempFile(t, stdout)
+	defer os.Remove(jsonTxFile.Name())
+
+	// Run the encode command, and trim the extras from the stdout capture
+	success, base64Encoded, _ := f.TxEncode(jsonTxFile.Name())
+	require.True(t, success)
+	trimmedBase64 := strings.Trim(base64Encoded, "\"\n")
+
+	// Run the decode command
+	success, stdout, stderr = f.TxDecode(trimmedBase64)
+	require.True(t, success)
+	require.Empty(t, stderr)
+
+	// Check that the transaction decodes as epxceted
+	var stdTx auth.StdTx
+	cdc := app.MakeCodec()
+	jsonTx := "{\"type\": \"auth/StdTx\",\"value\": " + stdout + "}"
```

Comment:
It is NOT elegant, but I don't know how to Unmarshal it. 
`json.Unmarshal(stdout, &stdTx)` also return error .

Reply:
Hold on, why can't you just decode() the encode()'s output? I'd expect something like what follows to work out of the box:

```
   sendTx() stdout                        encodeTx() stdout                      sendTx(stdout)
---------------------> [ encodeTx() ] -----------------------> [ decodeTx() ] -------------------->
```

Or am I getting something wrong?

Reply:
I got it !

`json.Unmarshal()` can not deal with `interface Msg`, and all the types and interfaces had been registed by `app.MakeCodec()`. 

So, `jsonTx := "{\"type\": \"auth/StdTx\",\"value\": " + stdout + "}"` is necessary.

![image](https://user-images.githubusercontent.com/7357939/56347956-08dd1100-61f8-11e9-94f3-0d015c911d36.png)


```go
package tx

import (
	"encoding/json"
	"fmt"
	"github.com/cosmos/cosmos-sdk/x/auth"
	"github.com/stretchr/testify/require"
	"testing"
)

var (
	stdoutWithoutType = `{
        "msg":[
            {
                "type":"cosmos-sdk/MsgSend",
                "value":{
                    "from_address":"cosmos1rpsdl0wdk42jjf9sp57d3eflm8q8ektuz66at3",
                    "to_address":"cosmos1e4lhppa79t40eqdgl5her60p4zqtenyj839y8z",
                    "amount":[
                        {
                            "denom":"muon",
                            "amount":"1000"
                        }
                    ]
                }
            }
        ],
        "fee":{
            "amount":[
                {
                    "denom":"muon",
                    "amount":"5000"
                }
            ],
            "gas":"200000"
        },
        "signatures":[
            {
                "pub_key":{
                    "type":"tendermint/PubKeySecp256k1",
                    "value":"A9dZ7e2azCHXp9hoTbXxTfiH2elBIet4Crj/4t5vecX0"
                },
                "signature":"TSNafIKca3+hEA48Fbf66gPJJscXkVvqactKHZOMPnBzB0PtNeHSFXWPl1lr1OR5p01V6eaz8zuJH7ukvTRGog=="
            }
        ],
        "memo":"userid001"
    }
`
)

func TestUnmarshalWithJson(t *testing.T) {
	var stdTx auth.StdTx
	err := json.Unmarshal([]byte(stdoutWithoutType), &stdTx)
	require.Nil(t, err)
	fmt.Println(stdTx)
}
```